src/webgpu/shader/validation/expression/call/builtin/smoothstep.spec.ts

```diff
@@ -0,0 +1,239 @@
+const builtin = 'smoothstep';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import {
+  Type,
+  concreteTypeOf,
+  elementTypeOf,
+  isFloatType,
+  kAllScalarsAndVectors,
+  kConvertableToFloatScalarsAndVectors,
+  scalarTypeOf,
+} from '../../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValuesTypes = objectsToRecord(kConvertableToFloatScalarsAndVectors);
+const kArgumentTypes = objectsToRecord(kAllScalarsAndVectors);
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() rejects invalid values
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValuesTypes))
+      .filter(u => stageSupportsType(u.stage, kValuesTypes[u.type]))
+      .beginSubcases()
+      .expand('value1', u => [-1000, -10, 0, 10, 1000])
+      .expand('value2', u => [-1000, -10, 0, 10, 1000])
+  )
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kValuesTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const type = kValuesTypes[t.params.type];
+
+    // We expect to fail if low == high as it results in a DBZ
+    const expectedResult = t.params.value1 !== t.params.value2;
+
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      expectedResult,
+      [type.create(t.params.value1), type.create(t.params.value2), type.create(0)],
+      t.params.stage,
+      /* returnType */ concreteTypeOf(type, [Type.abstractFloat, Type.f32, Type.f16])
```

Comment:
Should `Type.abstractFloat` be in here? That's not a concrete type. Don't you just want `Type.f32` for this test? That's the only thing `abstract-float` would convert to here, no?

Reply:
Hmm, yes, I suppose so. I was thinking of this arg more like what types are allowed for the builtin.

Reply:
Done.